tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Analyzer.scala

```diff
@@ -350,6 +352,124 @@ final class Analyzer(semantics: Semantics, outputMode: OutputMode,
       loop(this)
     }
 
+    def tryLookupReflProxyMethod(proxyName: String): Option[MethodInfo] = {
+      if (!allowAddingSyntheticMethods) {
+        tryLookupMethod(proxyName)
+      } else {
+        /* The lookup for a target method in this code implements the
+         * algorithm defining `java.lang.Class.getMethod`. This mimics how
+         * reflective calls are implemented on the JVM, at link time.
+         *
+         * Caveat: protected methods are not ignored. This can only make an
+         * otherwise invalid reflective call suddenly able to call a protected
+         * method. It never breaks valid reflective calls. This could be fixed
+         * if the IR retained the information that a method is protected.
+         */
+
+        @tailrec
+        def loop(ancestorInfo: ClassInfo): Option[MethodInfo] = {
+          if (ancestorInfo ne null) {
+            ancestorInfo.methodInfos.get(proxyName) match {
+              case Some(m) =>
+                assert(m.isReflProxy && !m.isAbstract)
+                Some(m)
+
+              case _ =>
+                ancestorInfo.findProxyMatch(proxyName) match {
+                  case Some(target) =>
+                    val targetName = target.encodedName
+                    Some(ancestorInfo.createReflProxy(proxyName, targetName))
+
+                  case None =>
+                    loop(ancestorInfo.superClass)
+                }
+            }
+          } else {
+            None
+          }
+        }
+
+        loop(this)
+      }
+    }
+
+    private def findProxyMatch(proxyName: String): Option[MethodInfo] = {
+      val candidates = methodInfos.valuesIterator.filter { m =>
+        // TODO In theory we should filter out protected methods
+        !m.isReflProxy && !m.isExported && !m.isAbstract &&
+        reflProxyMatches(m.encodedName, proxyName)
+      }.toSeq
+
+      val targets = candidates.filterNot { c =>
+        val resultType = methodResultType(c.encodedName)
+        candidates.exists { other =>
+          (other ne c) &&
+          isMoreSpecific(methodResultType(other.encodedName), resultType)
+        }
+      }
+
+      targets.headOption
```

Comment:
You should maybe retain the comment why and that this is just as broken than on the JVM (i.e. `targets` may contain multiple elements here, etc.)